File: 46.全排列.java
import java.util.LinkedList;
import java.util.List;

/*
 * @Description: 
 * @Author: yangyuxiang
 * @Date: 2021-04-21 07:17:16
 * @LastEditors: yangyuxiang
 * @LastEditTime: 2021-04-21 07:42:11
 * @FilePath: /leetcode/46.全排列.java
 */
/*
 * @lc app=leetcode.cn id=46 lang=java
 *
 * [46] 全排列
 */

// @lc code=start
class Solution {
    
    List<List<Integer>> paths = new LinkedList<>();
    
    public List<List<Integer>> permute(int[] nums) {
        
        LinkedList<Integer> path = new LinkedList<>();
        backtrace(nums, path);
        return paths;

    }
    
    public void backtrace(int[] nums, LinkedList<Integer> path) {
        if(nums.length == path.size()) {
            paths.add(new LinkedList(path));
            return;
        }
        for (int num : nums) {
            if (path.contains(num)) {
                continue;
            }
            path.add(num);
            backtrace(nums, path);
            path.removeLast();

        }
        
    }
}
// @lc code=end

File: 46.全排列.py
#!/usr/bin/env python
# coding=utf-8
'''
Description: 
Author: yangyuxiang
Date: 2021-04-20 22:48:13
LastEditors: yangyuxiang
LastEditTime: 2021-07-03 23:18:35
FilePath: /leetcode/46.全排列.py
'''
#
# @lc app=leetcode.cn id=46 lang=python
#
# [46] 全排列
#

# @lc code=start
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        paths = []
        visited = [False] * len(nums)

        def backtrace(path):
            if len(path) == len(nums):
                paths.append(path[:])
            for i in range(len(nums)):
                if visited[i] == False:
                    path.append(nums[i])
                    visited[i] = True
                    backtrace(path)
                    path.pop()
                    visited[i] = False
        backtrace([])
        return paths

if __name__ == '__main__':
    nums = [1,2,3]
    print(Solution().permute(nums))

        
# @lc code=end

